### PR TITLE
Doubly linked list

### LINKED ISSUE #187: Doubly linked list
There doesn't seem to be a doubly linked list here.
Can I add one?

### PR DESCRIPTION
**Pull request by korrat**
Added an implementation of a doubly linked list

- [x] Read the [contribution guidelines][contrib-guidelines]
- [x] Added [unit tests][unit-tests] (or unit tests have already been added)
- [x] Updated the [Contents][contents]

Closes #187

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md
[unit-tests]: https://github.com/faheel/Algos/blob/master/C%2B%2B/UNIT_TESTS.md
[contents]: https://github.com/faheel/Algos/tree/master/C%2B%2B#contents

### CODE REVIEW

**Comment by alxmjo** [comment]
Hi @Korrat. Thank you for your contribution. Please see the issues raised by the Codacy quality review above and update your code accordingly.

**Comment by korrat** [comment]
Sorry, I dropped the ball on this. Is there anything I should do to get this merged?

**Comment by faheel** [comment]
@korrat You need to move the header file to `include/data_structure/linked_list/doubly_linked_list.hpp` and the test file to `test/data_structure/linked_list/doubly_linked_list.cpp`. In the meanwhile, me an/or @alxmjo will review your code in detail.

**Comment by korrat** [comment]
Is this waiting on anything I should do?

**Comment by alxmjo** [comment]
Thanks for the reminder. I will check this out this weekend. In the meantime, can you correct the following spacing issue in your tests?

`REQUIRE( temp.length() == 0);` → `REQUIRE(temp.length() == 0);`

File: C++/include/data_structure/linked_list/doubly_linked_list.hpp [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+class DoublyLinkedList {
+    private:
+        Node<T> *head;
+        Node<T> *tail;
+        int size;
+
+    public:
+        DoublyLinkedList();
+        ~DoublyLinkedList();
+        bool is_empty(void);
```
Comment (alxmjo) [inline C++/include/data_structure/linked_list/doubly_linked_list.hpp:100]: Please test `is_empty()`

File: C++/include/data_structure/linked_list/doubly_linked_list.hpp [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        ~Node() {
+            this->prev = nullptr;
+            this->next = nullptr;
+        }
+
+        T get_value(void) {
+            return this->value;
+        }
+
+        void set_value(const T& other) {
```
Comment (alxmjo) [inline C++/include/data_structure/linked_list/doubly_linked_list.hpp]: Please test `set_value()`

File: C++/test/data_structure/linked_list/doubly_linked_list.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
+#include "data_structure/linked_list/doubly_linked_list.hpp"
```
Comment (alxmjo) [inline C++/test/data_structure/linked_list/doubly_linked_list.cpp:1]: Add test cases with negative numbers and numbers greater than 5.

**Review by alxmjo — CHANGES_REQUESTED** [review]
Looks good! See inline comments for a few minor changes, and please update the contents (`C++/README.md`) to reflect the addition.

File: C++/test/data_structure/linked_list/doubly_linked_list.cpp [hunk 4]
```diff
@@ -118,3 +118,18 @@ TEST_CASE("Removing values from rear", "[doubly-linked-list]") {
     REQUIRE(temp.length() == 0);
 }
 
+TEST_CASE("Inserting arbitray values", "[doubly-linked-list]") {
```
Comment (alxmjo) [inline C++/test/data_structure/linked_list/doubly_linked_list.cpp]: "arbitray" → "arbitrary"

**Review by alxmjo — CHANGES_REQUESTED** [review]
Thanks for the changes! Everything looks good to me. Just need to fix this one typo.

**Review by alxmjo — APPROVED** [review]
These look good to me. I'll wait a day or so to merge in case @faheel wants to take a look, too. After that I'll go ahead and merge. Thank you for contributing!
